application/utils/external_project_parsers/parsers/cheatsheet_categorizer.py

```
@@ -0,0 +1,350 @@
+
+    return sorted(found)
+
+
+def _validate_labels(labels) -> List[str]:
+    """
+    Filter an LLM-returned label list to only approved TAXONOMY entries.
+    Returns [] if nothing valid remains (caller should fall back).
+    """
+    if not isinstance(labels, list):
+        return []
+    valid = [l for l in labels if isinstance(l, str) and l in TAXONOMY]
```

Ruff (0.15.15)

[error] Ambiguous variable name: `l`

(E741)